groups-dev/script.js

```diff
@@ -0,0 +1,323 @@
+        const oldFilteredGroupsIds = obj.filteredGroupsIds;
+        obj[prop] = value;
+
+        if (oldFilteredGroupsIds == null) break;
+        // Optimization possible: Only rerender groups that have been added/removed
+        renderAllGroups();
+        break;
+      case 'search':
+        if (value === '') {
+          if (dataStore.groups == null) break;
+          dataStore.filteredGroupsIds = Object.values(dataStore.groups).map(
+            (group) => group.id,
+          );
+        } else {
+          const search = value.toLowerCase();
+          dataStore.filteredGroupsIds = Object.values(dataStore.groups)
+            .filter((group) => group.title.toLowerCase().includes(search))
+            .map((group) => group.id);
+        }
+        obj[prop] = value;
+        break;
+      case 'isGroupCreationModalOpen':
+        obj[prop] = value;
+
+        if (value) {
+          renderGroupCreationModal();
+        } else {
+          removeGroupCreationModal();
+        }
+        break;
+      case 'userSelf':
+        obj[prop] = value;
+        break;
+      case 'discordId':
+        obj[prop] = value;
+        break;
+      default:
+        throw new Error('Invalid property set');
+    }
+    return true;
+  },
+};
+
+const dataStore = new Proxy(
+  {
+    userSelf: null,
+    groups: null,
+    filteredGroupsIds: null,
+    search: '',
+    discordId: null,
+    isCreateGroupModalOpen: false,
+  },
+  handler,
+);
+
+// Lifecycle functions
+
+const onCreate = () => {
+  renderLoadingCards();
+  renderLoadingNavbarProfile();
+
+  getUserSelf()
+    .then((data) => {
+      if (data.statusCode === 401) {
+        renderNavbarProfileSignin();
+        renderNotAuthenticatedPage();
+        return;
+      } else if (data.error) {
+        throw new Error(data);
+      }
+      dataStore.userSelf = data;
+      afterAuthentication();
+    })
+    .catch((err) => {
+      console.error(err);
+      removeLoadingCards();
+    })
+    .finally(() => removeLoadingNavbarProfile());
+
+  bindSearchInput();
+  bindSearchFocus();
+  bindGroupCreationButton();
+};
+
+const afterAuthentication = async () => {
+  renderNavbarProfile();
+  Promise.all([getDiscordGroups(), getUserGroupRoles()])
+    .then(([groups, roleData]) => {
+      dataStore.filteredGroupsIds = groups.map((group) => group.id);
+      dataStore.groups = groups.reduce((acc, group) => {
+        let title = group.rolename
+          .replace('group-', '')
+          .split('-')
+          .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+          .join(' ');
+        acc[group.id] = {
+          id: group.id,
+          title: title,
+          count: group.memberCount,
+          isMember: group.isMember,
+          roleId: group.roleid,
+          description: group.description,
+          isUpdating: false,
+        };
+        return acc;
+      }, {});
+      dataStore.discordId = roleData.userId;
+    })
+    .then(() => removeLoadingCards());
+};
```

Comment:
smart, 
can we have a better name for this, instead of when this function be called.

Reply:
I tried to replicate life cycle methods. onCreate, afterAuthentication came from that. 

What name would you suggest?

Reply:
that is interesting, great then 👍
any blog on these lifecycle methods?